Documentation/DesignDocs/corewcfqueue-design.md

```diff
@@ -0,0 +1,190 @@
+# Objective 
+The objective of this document is to provide details on how to build a generic queue transport inside CoreWCF so that it can enable various queue binding. This document is based on the discussion outlined here https://github.com/CoreWCF/CoreWCF/issues/164 .
+
+# Why are we building
+When WCF was build there were not many queing technique and there was no concept of cloud. With multiple cloud providers supporting queing technique and many queing technique widely available, we wanted to build a generic transport, so that various queuing mechanism can be supported in CoreWCF.
+
+# Class Diagram
+![CoreWCF Generic Queue](/Documentation/DesignDocs/generic_queue_class_diagram.png?raw=true)
+
+# Class Details
+#### IQueueMiddlewareBuilder and QueueMiddlewareBuilder
+- IQueueMiddlewareBuilder provides signatures for 2 important methods, ```Use``` and ```Build```. 
+```csharp
+internal interface IQueueMiddlewareBuilder 
+{ 
+   IServiceProvider QueueServices { get; set; } 
+   IDictionary<string, object> Properties { get; } 
+   IQueueMiddlewareBuilder Use(Func<QueueMessageDispatcherDelegate, QueueMessageDispatcherDelegate> middleware); 
+   QueueMessageDispatcherDelegate Build(); 
+} 
+```
+- ```Use``` adds middleware delegates to the Queue pipeline 
+- ```Build``` builds the delegate used by this application to process requests received from the queue.
+#### QueueMiddleWareBuilderExtension
+- QueueMiddleWareBuilderExtension defines ```UseMiddleware``` method which specifically search for ```InvokeAsync``` method in each middleware class and construct it as delegate to be invoked by the middleware pipeline. This is similar to the NetTcp class UseMiddlewareFramingConnectionHandshakeExtensions and the ASP.NET Core UseMiddlewareExtensions..
+```csharp
+public class QueueMiddleWareBuilderExtension
+{
+    internal const string InvokeAsyncMethodName = "InvokeAsync";
+    public static IQueueMiddlewareBuilder UseMiddleware<TMiddleware>(this IQueueMiddlewareBuilder app, params object[] args)
+    {
+         return app.UseMiddleware(typeof(TMiddleware), args);
+    }
+    
+    public static IQueueMiddlewareBuilder UseMiddleware(this IQueueMiddlewareBuilder app, Type middleware, params object[] args)
+    {
+        //TODO
+     }
+    }
+```
+#### QueueServiceWebHostBuilderExtension
+- QueueServiceWebHostBuilderExtension is the WebHostBuilder extension which stiches all the middleware pipeline and ```QueuePollingService``` which will be called during service startup for queuing mechanism to work. e:g:- 
+```csharp
+internal sealed class Startup
+{
+    public void ConfigureServices(IServiceCollection services)
+    {
+      services.AddServiceModelServices();
+      services.UseCoreWCFQueue();
+ }
+ ```
+ - It first keeps an instance of IQueueMiddlewareBuilder for chaining all the middleware components to be used
```

Comment:
This isn't making a lot of sense. Do you mean it adds an implementation of IQueueMiddlewareBuilder to DI so that a Queue based binding is able to build the transport layer stack by adding middleware? It will add other needed components to DI such as QueueHandShakeMiddleWare and QueuePollingService which are necessary to build and run a Queue based service.  

We also need to rename QueueHandShakeMiddle. The NetTcp one is called that because there is a connection handshake which is completed in stages (here's my endpoint, here's my framing mode, here's the stream upgrade mime type I wish to upgrade etc). There is no handshake with Queue. One name that seems appropriately short but descriptive is (and I'm not the only person who can have an opinion on this, so feel free to suggest other names) QueueDispatchMiddleware? QueuePollingService seems like a good name, although QueuePollingHostedService might be a better fit as it describes what kind of service it is (it implements IHostedService).

Reply:
QueueFetchAndDispatchMiddleware ? => As it describes the fact that it first fetches the queue by using the transport and dispatch service dispatcher to process.
QueuePollingHostedService would be better as it clearly speaks about the IHostedService